File: swift/Sources/TokenTrackerCore/ScanScheduler.swift
//
//  ScanScheduler.swift
//  TokenTrackerCore
//
//  移植自 server.py 的 ScanService：一次只跑一个扫描，HTTP 动作与
//  桌面定时器共用锁。时钟/等待/线程可注入（测试隔离）；失败记录时间戳
//  并释放锁；stop 不取消进行中的事务，只阻止新扫描。
//

import Foundation

/// 可注入的线程抽象（测试用 InlineScanThread 同步执行）。
public protocol ScanThreadHandle {
    func start()
    func join(timeout: TimeInterval)
    var isAlive: Bool { get }
    var isCurrentThread: Bool { get }
}

public final class RealScanThread: ScanThreadHandle, @unchecked Sendable {
    /// 独立状态盒：避免 init 里 Thread(block:) 捕获未初始化完的 self。
    private final class State: @unchecked Sendable {
        var alive = true
        let lock = NSLock()
    }
    private let state = State()
    private let thread: Thread

    public init(_ block: @escaping @Sendable () -> Void) {
        let state = self.state
        thread = Thread(block: {
            block()
            state.lock.lock()
            state.alive = false
            state.lock.unlock()
        })
    }

    public func start() { thread.start() }

    public func join(timeout: TimeInterval) {
        let deadline = Date().addingTimeInterval(timeout)
        while isAlive && Date() < deadline {
            Thread.sleep(forTimeInterval: 0.01)
        }
    }

    public var isAlive: Bool {
        state.lock.lock()
        defer { state.lock.unlock() }
        return state.alive && thread.isExecuting && !thread.isFinished
    }

    public var isCurrentThread: Bool { Thread.current == thread }
}

public struct ScanSchedulerStatus: Equatable, Sendable {
    public struct Last: Equatable, Sendable {
        public var startedAt: Double = 0
        public var finishedAt: Double = 0
        public var source: String = ""
        public var done = false
        public var error: String?
        public var added: Int = 0
        public var repriced: Int = 0
        public var counterResets: Int = 0
        public var warnings: [String] = []
    }
    public var running = false
    public var last: Last?

    public init(running: Bool = false, last: Last? = nil) {
        self.running = running
        self.last = last
    }
}

public final class ScanScheduler: @unchecked Sendable {
    public typealias ScanBlock = @Sendable (_ tools: [String]?, _ full: Bool) throws
        -> (results: [String: ScanOutcome], repriced: Int)
    public typealias ThreadFactory = (@Sendable @escaping () -> Void) -> ScanThreadHandle

    private let scan: ScanBlock
    private let interval: Double
    private let clock: () -> Double
    private let waitImpl: (Double) -> Bool   // true = 收到停止信号（对齐 Event.wait）
    private let threadFactory: ThreadFactory

    private let lock = NSLock()
    private final class StopFlag: @unchecked Sendable {
        var value = false
    }
    private let stopFlag = StopFlag()
    private var status = ScanSchedulerStatus()
    private var worker: ScanThreadHandle?
    private var timer: ScanThreadHandle?

    private var isStopped: Bool {
        lock.lock()
        defer { lock.unlock() }
        return stopFlag.value
    }

    /// 扫描完成回调（调用方负责切主线程）。
    public var onFinish: (() -> Void)?

    public init(scan: @escaping ScanBlock, interval: Double = 60,
                clock: (() -> Double)? = nil,
                wait: ((Double) -> Bool)? = nil,
                threadFactory: ThreadFactory? = nil) {
        self.scan = scan
        self.interval = interval
        let flag = stopFlag
        let lockRef = lock
        let stopCheck: @Sendable () -> Bool = {
            lockRef.lock()
            defer { lockRef.unlock() }
            return flag.value
        }
        self.clock = clock ?? { Date().timeIntervalSince1970 }
        self.waitImpl = wait ?? { timeout in
            let deadline = Date().addingTimeInterval(timeout)
            while Date() < deadline {
                if stopCheck() { return true }
                Thread.sleep(forTimeInterval: min(0.05, max(0, deadline.timeIntervalSinceNow)))
            }
            return stopCheck()
        }
        self.threadFactory = threadFactory ?? { RealScanThread($0) }
    }

    public func snapshot() -> ScanSchedulerStatus {
        lock.lock()
        defer { lock.unlock() }
        return status
    }

    /// 请求一次扫描；运行中/已停止返回 false。
    @discardableResult
    public func request(tools: [String]? = nil, full: Bool = false, source: String = "manual") -> Bool {
        lock.lock()
        if stopFlag.value || status.running {
            lock.unlock()
            return false
        }
        status.running = true
        status.last = ScanSchedulerStatus.Last(startedAt: clock(), source: source)
        let worker = threadFactory { [weak self] in self?.run(tools: tools, full: full) }
        self.worker = worker
        lock.unlock()
        worker.start()
        return true
    }

    private func run(tools: [String]?, full: Bool) {
        var scanError: String?
        var added = 0, repriced = 0, resets = 0, warnings: [String] = []
        do {
            let result = try scan(tools, full)
            repriced = result.repriced
            var errors: [String] = []
            for (name, outcome) in result.results {
                if let e = outcome.error { errors.append("\(name): \(e)") }
                if let w = outcome.warning { warnings.append(w) }
                added += outcome.added
                resets += outcome.counterResets
            }
            if !errors.isEmpty { scanError = errors.joined(separator: "; ") }
        } catch {
            // 失败的扫描不得污染共享锁
            scanError = String(describing: error)
        }
        lock.lock()
        status.last?.done = true
        status.last?.finishedAt = clock()
        status.last?.error = scanError
        status.last?.added = added
        status.last?.repriced = repriced
        status.last?.counterResets = resets
        status.last?.warnings = warnings
        status.running = false
        lock.unlock()
        onFinish?()
    }

    /// 启动自动调度：立即扫一次，此后每 interval 秒增量扫描。
    public func startAuto() {
        lock.lock()
        if stopFlag.value || timer != nil {
            lock.unlock()
            return
        }
        let timer = threadFactory { [weak self] in self?.autoLoop() }
        self.timer = timer
        lock.unlock()
        timer.start()
    }

    private func autoLoop() {
        request(source: "automatic")
        while !waitImpl(interval) {
            if isStopped { break }
            request(source: "automatic")
        }
    }

    /// 停止调度：不取消进行中的 SQLite 事务；阻止新扫描并有界等待。
    public func stop() {
        lock.lock()
        stopFlag.value = true
        let worker = self.worker
        let timer = self.timer
        lock.unlock()
        for thread in [worker, timer] {
            if let thread, !thread.isCurrentThread { thread.join(timeout: 2) }
        }
    }
}

File: windows/src/TokenTracker.Core/Scheduling/ScanScheduler.cs
using TokenTracker.Core.Scanners;

namespace TokenTracker.Core.Scheduling;

public sealed record ScanLast
{
    public double StartedAt { get; init; }
    public double FinishedAt { get; init; }
    public string Source { get; init; } = "";
    public bool Done { get; init; }
    public string? Error { get; init; }
    public int Added { get; init; }
    public int ActivityAdded { get; init; }
    public int ActivityUpdated { get; init; }
    public int Repriced { get; init; }
    public int CounterResets { get; init; }
    public IReadOnlyList<string> Warnings { get; init; } = [];
}

public sealed record ScanSchedulerStatus(bool Running, ScanLast? Last);

/// <summary>
/// 移植 ScanScheduler.swift（server.py 的 ScanService）：一次只跑一个扫描，手动与定时共用；
/// 失败记录后释放；Stop 不取消进行中的事务，只阻止新扫描并有界等待。
/// </summary>
public sealed class ScanScheduler
{
    public delegate (Dictionary<string, ScanOutcome> Results, int Repriced) ScanFunc(IReadOnlyList<string>? tools, bool full);

    readonly ScanFunc _scan;
    readonly Func<double> _clock;
    readonly object _lock = new();
    readonly ManualResetEventSlim _stopSignal = new(false);
    double _interval;
    bool _stopped;
    ScanSchedulerStatus _status = new(false, null);
    Thread? _worker;
    Thread? _timer;

    /// <summary>扫描完成回调（在扫描线程上调用，调用方负责切 UI 线程）。</summary>
    public event Action? Finished;

    public ScanScheduler(ScanFunc scan, double interval = 60, Func<double>? clock = null)
    {
        _scan = scan;
        _interval = interval;
        _clock = clock ?? (() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0);
    }

    public ScanSchedulerStatus Snapshot()
    {
        lock (_lock) return _status;
    }

    public double IntervalSeconds
    {
        get
        {
            lock (_lock) return _interval;
        }
    }

    /// <summary>运行中热改间隔：自动循环每轮等待前读最新值。</summary>
    public void SetInterval(double seconds)
    {
        if (seconds <= 0) return;
        lock (_lock) _interval = seconds;
    }

    /// <summary>请求一次扫描；运行中/已停止返回 false。</summary>
    public bool Request(IReadOnlyList<string>? tools = null, bool full = false, string source = "manual")
    {
        Thread worker;
        lock (_lock)
        {
            if (_stopped || _status.Running) return false;
            _status = new ScanSchedulerStatus(true, new ScanLast { StartedAt = _clock(), Source = source });
            worker = new Thread(() => Run(tools, full)) { IsBackground = true, Name = "tt-scan", Priority = ThreadPriority.BelowNormal };
            _worker = worker;
        }
        worker.Start();
        return true;
    }

    void Run(IReadOnlyList<string>? tools, bool full)
    {
        string? scanError = null;
        int added = 0, activityAdded = 0, activityUpdated = 0, repriced = 0, resets = 0;
        var warnings = new List<string>();
        try
        {
            var (results, rep) = _scan(tools, full);
            repriced = rep;
            var errors = new List<string>();
            foreach (var (name, outcome) in results)
            {
                if (outcome.Error is not null) errors.Add($"{name}: {outcome.Error}");
                if (outcome.Warning is not null) warnings.Add(outcome.Warning);
                added += outcome.Added;
                activityAdded += outcome.ActivityAdded;
                activityUpdated += outcome.ActivityUpdated;
                resets += outcome.CounterResets;
            }
            if (errors.Count > 0) scanError = string.Join("; ", errors);
        }
        catch (Exception e)
        {
            // 失败的扫描不得污染共享锁
            scanError = e.Message;
        }
        lock (_lock)
        {
            _status = new ScanSchedulerStatus(false, (_status.Last ?? new ScanLast()) with
            {
                Done = true, FinishedAt = _clock(), Error = scanError, Added = added, ActivityAdded = activityAdded,
                ActivityUpdated = activityUpdated, Repriced = repriced, CounterResets = resets, Warnings = warnings,
            });
        }
        Finished?.Invoke();
    }

    /// <summary>启动自动调度：立即扫一次，此后每 interval 秒增量扫描。</summary>
    public void StartAuto()
    {
        Thread timer;
        lock (_lock)
        {
            if (_stopped || _timer is not null) return;
            timer = new Thread(AutoLoop) { IsBackground = true, Name = "tt-scan-timer" };
            _timer = timer;
        }
        timer.Start();
    }

    void AutoLoop()
    {
        Request(source: "automatic");
        while (!_stopSignal.Wait(TimeSpan.FromSeconds(IntervalSeconds)))
            Request(source: "automatic");
    }

    /// <summary>停止调度：不取消进行中的 SQLite 事务；阻止新扫描并有界等待。</summary>
    public void Stop()
    {
        Thread? worker, timer;
        lock (_lock)
        {
            _stopped = true;
            worker = _worker;
            timer = _timer;
        }
        _stopSignal.Set();
        foreach (var thread in new[] { worker, timer })
            if (thread is not null && thread != Thread.CurrentThread) thread.Join(TimeSpan.FromSeconds(2));
    }
}
